### ISSUE TITLE
Set the "test" compiler flag automatically when running crystal spec

### ISSUE THREAD

**Issue by lipanski** [crystal-lang/crystal#5783]
This is a feature request.

I find compiler flags quite useful in Crystal and they can solve problems like https://github.com/crystal-lang/crystal/issues/5220 in similar ways the proposed `CRYSTAL_ENV` environment variable would.

Crystal already sets the `release` compiler flag automatically whenever building with `--release`.

I suggest setting the `test` (or `spec`?) compiler flag automatically whenever running `crystal spec`. This can help in application and shard development to exclude/include test-specific behaviour in a **conistent**, Crystal-wide manner. Shard developers would be able to rely on the presence of this flag whenever shard users are running specs. Application developers wouldn't have to add `-Dtest` explicitly in order to use this flag while running their own tests.

---

I can help with the development, if this feature request is something of interest.

**Comment by RX14** [comment]
I'm not sure it's a good idea to encourage libraries to start altering their configuration just for specs.

**Comment by lipanski** [comment]
> altering their configuration just for specs

you do need this from time to time. I'm not say *encourage* shard developers (or application developers) to do this but when they need it, they could rely on a common behaviour.

my use case is based on running integration tests with https://github.com/amberframework/garnet-spec on top of a web app. so far there's just one thing I need to warp inside `flag?(:test)` in my application code, but the compiler flag proved useful and I thought it's a nice, clean solution that could be useful to others as well.

**Comment by ysbaddaden** [comment]
We set the `release`, `static` and `debug` flags because the compiler knows about them: they have a meaning and are used. Introducing `test` would come out of the blue: it means nothing to the compiler.

Specifying `-Dtest` ain't hard. Adding some runtime code —wrapped in `unless flag(:release)`— to check for a `test` cookie is another solution, and so on.

**Comment by RX14** [comment]
If you want to spec an application, have you tried moving the top-level code into a method, then creating a `src/entrypoint.cr` file containing similar to:
```cr
require "./myapp"
MyApp.run
```

This is what I do to be able to spec my application code. Libraries obviously shouldn't have this issue.

**Comment by lipanski** [comment]
> creating a `src/entrypoint.cr` file 

that's pretty much where I started from. I didn't like this (having to introduce an entrypoint just because of tests), so I got rid of the entrypoint file and went with:

```crystal
{% unless flag?(:test) %}
MyApp.run
{% end %}
```

...and ran my tests with `crystal spec -Dtest`. and then I thought it could be generally convenient to have the `test` flag set implicitly, pretty much like when I'd introduce release-specific logic and wrap it inside `flag?(:release)`.

this is strictly about ergonomics and doesn't make any *major* difference, but I still wanted to see if someone would be interested in this workflow.
